src/writers/s3.sink.cpp

```diff
@@ -0,0 +1,162 @@
+                     std::string_view object_key,
+                     std::shared_ptr<common::S3ConnectionPool> connection_pool)
+  : bucket_name_{ bucket_name }
+  , object_key_{ object_key }
+  , connection_pool_{ connection_pool }
+{
+    CHECK(!bucket_name_.empty());
+    CHECK(!object_key_.empty());
+    CHECK(connection_pool_);
+}
+
+zarr::S3Sink::~S3Sink()
+{
+    if (!is_multi_part_upload_() && n_bytes_buffered_ > 0) {
+        if (!put_object_()) {
+            LOGE("Failed to upload object: %s", object_key_.c_str());
+        }
+    } else if (is_multi_part_upload_()) {
+        if (n_bytes_buffered_ > 0 && !flush_part_()) {
+            LOGE("Failed to upload part %zu of object %s",
+                 parts_.size() + 1,
+                 object_key_.c_str());
+        }
+        if (!finalize_multipart_upload_()) {
+            LOGE("Failed to finalize multipart upload of object %s",
+                 object_key_.c_str());
+        }
+    }
+}
+
+bool
+zarr::S3Sink::write(size_t offset, const uint8_t* data, size_t bytes_of_data)
+{
+    CHECK(data);
+    CHECK(bytes_of_data);
+
+    while (bytes_of_data > 0) {
+        const auto bytes_to_write =
+          std::min(bytes_of_data, part_buffer_.size() - n_bytes_buffered_);
+
+        if (bytes_to_write) {
+            std::copy_n(
+              data, bytes_to_write, part_buffer_.begin() + n_bytes_buffered_);
+            n_bytes_buffered_ += bytes_to_write;
+            data += bytes_to_write;
+            bytes_of_data -= bytes_to_write;
+        }
+
+        if (n_bytes_buffered_ == part_buffer_.size() && !flush_part_()) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool
+zarr::S3Sink::put_object_()
+{
+    if (n_bytes_buffered_ == 0) {
+        return false;
+    }
+
+    auto connection = connection_pool_->get_connection();
+
+    bool retval = false;
+    try {
+        std::string etag =
+          connection->put_object(bucket_name_,
+                                 object_key_,
+                                 { part_buffer_.data(), n_bytes_buffered_ });
+        EXPECT(
+          !etag.empty(), "Failed to upload object: %s", object_key_.c_str());
+
+        retval = true;
+    } catch (const std::exception& exc) {
+        LOGE("Error: %s", exc.what());
+    }
+
+    // cleanup
+    connection_pool_->return_connection(std::move(connection));
+    n_bytes_buffered_ = 0;
+
+    return retval;
+}
+
+bool
+zarr::S3Sink::is_multi_part_upload_() const
+{
+    return !upload_id_.empty() && !parts_.empty();
+}
+
+bool
+zarr::S3Sink::flush_part_()
+{
+    if (n_bytes_buffered_ == 0) {
+        return false;
+    }
+
+    auto connection = connection_pool_->get_connection();
+
+    auto get_upload_id = [this, &connection]() -> std::string {
+        if (upload_id_.empty()) {
+            upload_id_ =
+              connection->create_multipart_object(bucket_name_, object_key_);
+        }
+        return upload_id_;
+    };
```

Comment:
I think this should be a separate method. Sorry if my comment was confusing—I typically refer to accessor methods as computed properties.